lib/propolis/src/cpuid.rs

```diff
@@ -0,0 +1,413 @@
+                ent.ebx &= !0xff000000;
+                ent.ebx |= ((*vcpuid as u32) & 0xff) << 24;
+            }
+        }
+
+        // logical CPU count (if SMT is enabled)
+        if let Some(num_vcpu) = self.num_vcpu.as_ref() {
+            if self.has_smt {
+                if let Some(ent) = set.get_mut(Ident(0x1, None)) {
+                    ent.edx |= (0x1 << 28);
+                    // bits 23:16 contain max IDs for logical CPUs in package
+                    ent.ebx &= !0xff0000;
+                    ent.ebx |= (num_vcpu.get() as u32) << 16;
+                }
+            }
+        }
+
+        set
+    }
+
+    fn fix_amd_cache_topo(&self, set: &mut Set) {
+        assert!(self.do_cache_topo);
+        let num = self.num_vcpu.unwrap().get();
+        for ecx in 0..u32::MAX {
+            match set.get_mut(Ident(0x8000001d, Some(ecx))) {
+                None => break,
+                Some(vals) => {
+                    // bits 7:5 hold the cache level
+                    let visible_count = match (vals.eax & 0b11100000 >> 5) {
+                        0b001 | 0b010 => {
+                            // L1/L2 shared by SMT siblings
+                            if self.has_smt {
+                                2
+                            } else {
+                                1
+                            }
+                        }
+                        0b011 => {
+                            // L3 shared by all vCPUs
+                            // TODO: segregate by sockets, if configured
+                            num as u32
+                        }
+                        _ => {
+                            // unceremonious handling of unexpected cache levels
+                            panic!("unsupported cache level");
+                        }
+                    };
+                    // the number of logical CPUs (minus 1) sharing this cache
+                    // is stored in bits 25:14
+                    vals.eax &= !(0xfff << 14);
+                    vals.eax |= (visible_count - 1) << 14;
+                }
+            }
+        }
+    }
+    fn fix_cpu_topo(&self, set: &mut Set) {
+        for topo in enum_iterator::all::<TopoKind>() {
+            // Nuke any existing info in order to potentially override it
+            let leaf = topo as u32;
+            set.remove_all(leaf);
+
+            let num_vcpu = match self.num_vcpu {
+                Some(n) => n.get() as u32,
+                None => {
+                    // Skip adding topo info if we lack vcpu count
+                    continue;
+                }
+            };
+
+            if self.cpu_topo.contains(&topo) {
+                match topo {
+                    TopoKind::StdB => {
+                        // Queries with invalid ecx will get all-zeroes
+                        set.insert(Ident(leaf, None), Entry::zero());
+                        if self.has_smt {
```

Comment:
Related to the above: if I want to test a guest OS with a zany set of outputs from a topology leaf, and I write a config file accordingly, and Propolis runs this logic, my settings will be silently overwritten. Should this return an error instead?

Reply:
I've tweaked the leaf clearing behavior.  LMK if that's fine for now.